keps/sig-cloud-provider/20190215-promoting-cloud-provider-labels.md

```diff
@@ -0,0 +1,178 @@
+These labels contain cloud provider information that may be critical to some advanced features (mainly scheduling).
+When these labels were first introduced, they were prefixed with "beta" as the maturity and usage of these labels were not known at the time.
+
+Today, the cloud provider specific labels are:
+* `beta.kubernetes.io/instance-type`
+* `failure-domain.beta.kubernetes.io/zone`
+* `failure-domain.beta.kubernetes.io/region`
+
+This KEP proposes to remove the beta labels eventually and replace them with their GA equivalents:
+* `kubernetes.io/instance-type`
+* `failure-domain.kubernetes.io/zone`
+* `failure-domain.kubernetes.io/region`
+
+## Motivation
+
+The labels mentioned above are consumed by almost all Kubernetes clusters that have cloud providers enabled. Given their maturity and widespread use, we should
+promote these labels from beta to GA.
+
+### Goals
+
+* promote cloud provider node/volume labels to GA with minimal visible changes to users.
+* remove the usage of "beta" cloud provider node/volume labels without breaking compatiblity guaranetees. This will span many Kubernetes versions as per the Kubernetes deprecation policy.
+
+### Non-Goals
+
+* introducing more labels
+* changing the behaviour of these labels within the Kubernetes system.
+
+## Proposal
+
+In order to promote these labels to GA safely, there will be a period in which both the "beta" and "GA" labels are applied to node and volume objects.
+This is required in order to maintain backwards compatiblity as many clusters rely on the beta labels today.
+
+Both new and existing resources will require both the beta and GA versions of the labels. For the case of existing resources, keeping the beta labels is a requirement in order for
+existing workloads to behave as expected. For the case of new resources, both labels are still requried as workloads may still consume the beta labels in some other resource
+that was not updated yet to use the GA labels. One possible example is where a deployment may still use the beta zone label (`failure-domain.beta.kubernetes.io/zone`) as a
+nodeSelector and not applying the beta labels to new nodes would mean new nodes in that zone would not be considered when pods are being scheduled.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Here is a break down of the implementation steps:
+
+1) [v1.14] update components to apply both the GA and beta labels to nodes & volumes.
+2) [v1.14] deprecate the beta labels.
+3) [v1.14] update the appropriate release notes & documentation to promote the use of GA labels over beta labels.
+4) [v1.15] continue to promote usage of GA labels over beta labels.
+5) [v1.16] continue to promote usage of GA labels over beta labels.
+6) [v1.17] continue to promote usage of GA labels over beta labels.
+7) [v1.18] stop applying beta labels to new resources, existing resources will continue to have those labels unless manually removed.
```

Comment:
Is there a timeframe for when the controllers can stop handling beta labels? If so, do we expect that users will need to manually update existing objects update/add the GA label, or will this be automated?  I'm running into a similar issue trying to [deprecate and remove the VolumeZone scheduling predicate](https://github.com/kubernetes/kubernetes/issues/72139)

Reply:
For this KEP I was thinking by v1.18 we should expect users to have removed the beta labels so it would be okay to remove logic for it in the controllers as well (I'll make this clearer).  I think it would be safer to not have any automation in place for removing labels though as users can be using those labels in their custom tooling/controllers.

Reply:
Sorry, I didn't mean automation for removing beta labels. I meant automation for adding GA labels to existing objects with the beta label.

@thockin do these labels apply to the api deprecation policy? It mentions annotations but not labels.

Reply:
Ahh I see sorry, yes I think we need to update kubelet/controllers to automatically add the GA labels to existing objects.